pkg/controller/azure/service/actuator.go

```diff
@@ -145,34 +154,33 @@ func (a *actuator) Delete(ctx context.Context, obj runtime.Object) error {
 	return nil
 }
 
+// ShouldFinalize returns true if the object should be finalized.
+func (a *actuator) ShouldFinalize(_ context.Context, obj runtime.Object) (bool, error) {
+	// Cast object to Service
+	var svc *corev1.Service
+	var ok bool
+	if svc, ok = obj.(*corev1.Service); !ok {
+		return false, errors.New("reconciled object is not a service")
+	}
+
+	// Return true if there are LoadBalancer IPs and the service should not be ignored
+	return len(getServiceLoadBalancerIPs(svc)) > 0 && !shouldIgnoreService(svc), nil
+}
+
 func getServiceLoadBalancerIPs(svc *corev1.Service) map[string]bool {
 	ips := make(map[string]bool)
-	if svc.Spec.Type == corev1.ServiceTypeLoadBalancer {
```

Comment:
It's not entirely wrong, but it's redundant. You are interested in the load balancer IPs, not in whether the service is of type "load balancer" or not. The real impact of the change is when the controller becomes aware of the change in the service type:
* without the change, it sees LoadBalancer -> ClusterIP immediately (before the IP is released), but ClusterIP -> LoadBalancer with a delay (only after the IP is allocated).
* with the change, it sees both changes with a delay (only after the IP is released / allocated).
I prefer the second since it's a bit more consistent and easier to reason about.